MWI: Minimal bound-keypair joining implementation

timothyb89 opened the pull request:
This includes a minimal implementation of bound-keypair joining. This first iteration requires preregistered public keys, and requires `unlimited` and `insecure` flags to be set on bound keypair tokens.

Minimal client-side implementation will be in a follow up PR.

RFD: #52546
Protos PR: https://github.com/gravitational/teleport/pull/53566
Client PR: https://github.com/gravitational/teleport/pull/54372

Closes #53373

timothyb89 committed:
MWI: Minimal bound-keypair joining implementation

This includes a minimal implementation of bound-keypair joining. This
first iteration requires preregistered public keys, and requires
`unlimited` and `insecure` flags to be set on bound keypair tokens.

Minimal client-side implementation will be in a follow up PR.

RFD: #52546
Closes #53373

timothyb89 commented:
Outstanding TODOs:

- [ ] Docs & misc cleanup
- [ ] Tests

timothyb89 committed:
Refactor challenge response function, rebase on updated protos branch

This includes a number of changes:
- Rebases on the latest protos branch. This includes removal of the
  new keypair field on initial join, and adds messages for
  interactive keypair rotation.
- Per the rebase, remaining_joins is removed in favor of using
  join_count for all calculations. The registration method and
  validatity checks have been updated to reference that instead.
- Refactors challenge response function to allow for keypair
  rotation. We still don't implement rotation but the handler now
  receives the full proto message and produces a full proto response,
  so that we can easily handle the rotation case in the future.
- Challenge validation checks time fields explicitly to ensure the
  client didn't tamper with them.
- Added some missing docstrings

timothyb89 committed:
Add joinserver test

timothyb89 committed:
Fix lint error and add docstring

timothyb89 committed:
Add tests for bound keypair challenge validation

timothyb89 committed:
Remove client side package intended for other PR

timothyb89 committed:
Fix various lints

timothyb89 committed:
Add tests for RegisterUsingBoundKeypairMethod()

timothyb89 committed:
Fix lints

timothyb89 commented on lib/auth/join_bound_keypair.go:
```diff
@@ -0,0 +1,483 @@
+		return trace.Wrap(err)
+	}
+
+	// TODO: Is this wise? End users _shouldn't_ modify this, but this could
+	// interfere with cluster backup/restore. Options seem to be:
+	// - Let users create/update with status fields. They can break things, but
+	//   maybe that's okay. No backup/restore implications.
+	// - Ignore status fields during creation and update. Any set value will be
+	//   discarded here, and during update. This would still have consequences
+	//   during cluster restores, but wouldn't raise errors, and the status
+	//   field would otherwise be protected from easy tampering. Users might be
+	//   confused as no user-visible errors would be raised if they used
+	//   `tctl edit`.
+	// - Raise an error if status fields are changed. Worst restore
+	//   implications, but tampering won't be easy, and will have some UX.
+	if tokenV2.Status.BoundKeypair != nil {
+		return trace.BadParameter("cannot create a bound_keypair token with set status")
+	}
```
Curious for thoughts on this. I think the simplest option is probably to just let users modify the resource freely, even if that has consequences.

boxofrad replied:
I think this is one of the reason a resource's "status" is actually a sub-resource (with different RBAC permissions) in Kubernetes.

timothyb89 replied:
Hmm, that makes sense. I'll just leave a note and keep it simple - we'll tell users not to modify the status field, but there won't be any actual guardrails in place so we don't get in the way of restore attempts.

timothyb89 committed:
Add basic provisioning token CheckAndSetDefaults() tests

timothyb89 committed:
Include bound public key in RegisterUsingBoundKeypairMethod return

This is passed back to clients as part of the proto certs message as
confirmation that rotation succeeded, so the value needed to be
plumbed through.

strideynet commented on api/types/provisioning.go:
```diff
@@ -80,6 +80,9 @@ const (
 	JoinMethodOracle JoinMethod = "oracle"
+	// JoinMethodBoundKeypair indicates the node will join using the Bound
+	// Keypair join method. See lib/boundkeypair for more.
+	JoinMethodBoundKeypair JoinMethod = "bound_keypair"
```
I went with underscore in the Azure Devops stuff I've been planning since I figured matching the join method to the spec. field was more consistent.

timothyb89 replied:
Hmm, fair enough, I'll stick with the underscore as-is, then. Thanks!

boxofrad commented on lib/boundkeypair/bound_keypair.go:
```diff
@@ -0,0 +1,150 @@
+
+// ChallengeDocument is a bound keypair challenge document. These documents are
+// sent in JSON form to clients attempting to authenticate, and are expected to
+// be sent back signed with a known public key.
```
Sorry to nitpick! Feel free to re-word too.

```suggestion
// be sent back signed with the private counterpart of a known public key.
```

boxofrad commented on lib/auth/join_bound_keypair.go:
```diff
@@ -0,0 +1,489 @@
+	case hasBoundPublicKey && hasBoundBotInstance && !hasIncomingBotInstance:
+		// Hard rejoin case, the client identity expired and a new bot instance
+		// is required. Consumes a rejoin.
+		if !spec.Joining.Unlimited && !hasJoinsRemaining {
+			return nil, "", trace.AccessDenied("no rejoins remaining")
+		}
```
Should this check actually be inside the function passed to `PatchToken` in case the join counter has been modified concurrently?

timothyb89 replied:
The decision is made here, but the mutator function (`mutateStatusConsumeJoin`) checks that:
- the join (now recovery) counter has not changed since the decision was made
- the join (now recovery) limit is at least the value it was when the decision was made

The mutator functions are run inside `PatchToken`, so those checks will always need to pass for the patch to succeed.

boxofrad reviewed (approved):
This looks great! Nice work 👏🏻

timothyb89 committed:
Fixes after upstream proto change

We renamed and tweaked a number of proto fields, so this updates
field references.

timothyb89 committed:
Apply suggestions from code review

Co-authored-by: Dan Upton <daniel.upton@goteleport.com>

timothyb89 committed:
Remove TODO

timothyb89 committed:
Merge branch 'master' into timothyb89/bound-keypair-minimal-impl

timothyb89 committed:
Fix missed field rename

timothyb89 committed:
Merge remote-tracking branch 'origin/timothyb89/bound-keypair-minimal-impl' into timothyb89/bound-keypair-minimal-impl

timothyb89 committed:
Fix broken test

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/bound-keypair-minimal-impl

timothyb89 commented:
Small update note: I've split out `lib/boundkeypair` into its own PR in #54766. That was dropped from this PR, and a few changes were requested there, so e.g. the experiment package was renamed here. There were no other functional changes.

timothyb89 committed:
Fix lurking nil pointer deref after field rename
